client/network/sync/src/warp.rs

```diff
@@ -77,8 +82,32 @@ where
 			set_id: 0,
 			authorities: warp_sync_provider.current_authorities(),
 			last_hash,
+			warp_sync_provider,
 		};
-		Self { client, warp_sync_provider, phase, total_proof_bytes: 0 }
+		Self { client, phase, total_proof_bytes: 0 }
+	}
+
+	///  Create a new instance, skip the proof downloading and verification and go directly to
+	/// target block
+	fn new_with_target_block(client: Arc<Client>, target_block: <B>::Header) -> Self {
+		let phase = Phase::TargetBlock(target_block);
+		Self { client, phase, total_proof_bytes: 0 }
+	}
+
+	///  Poll and wait for target block
+	pub fn poll_target_block(
+		client: Arc<Client>,
+		warp_sync_params: Option<&mut WarpSyncParams<B>>,
+		cx: &mut std::task::Context,
+	) -> Poll<Self> {
+		if let Some(WarpSyncParams::WaitForTarget(target_block)) = warp_sync_params {
+			return match target_block.poll_unpin(cx) {
+				Poll::Ready(Ok(target_block)) =>
+					Poll::Ready(WarpSync::new_with_target_block(client.clone(), target_block)),
+				_ => Poll::Pending,
+			}
+		}
+		Poll::Pending
 	}
```

Comment:
This should be some internals of `WarpSync`. There should be one method `new` that takes the `WarpSyncParams`. No one outside of warp sync should care what these parameters are and what they are doing.

Reply:
Sure I can do that, but we still need to pass `cx` so we can poll inside new right? Because we are calling `new` from the existing `poll` method?

Reply:
@bkchr I'd argue that waiting for the signal should be on the `ChainSync` level. `ChainSync` interprets the configuration and decides when the process should actually start, I.e. waiting for sufficient peers and in this case also waiting for the target block to be set.  WarpSync` is managing the warp sync after it has actually started.

Reply:
I mean I know that this currently is the case, but actually we want to have in the future all these things separated. Currently all of the logic is all over the place. We should try to add new things already to the place where they belong. I also know that just giving it the header is not really warp syncing.

Reply:
Got it now, I added a new phase `PendingTargetBlock` and we always just call `WarpSync::new`

I also added an if statement to only call `new` in poll when its a `WaitForTarget` or its awaiting a target block to ensure we do not break the existing functionality in `ChainSync` which waits for `MIN_PEERS_TO_START_WARP_SYNC`